tests/posix/common/src/nanosleep.c

```diff
@@ -143,7 +143,7 @@ static void common_lower_bound_check(int selection, clockid_t clock_id, int flag
 	zassert_equal(rem.tv_sec, 0, "actual: %d expected: %d", rem.tv_sec, 0);
 	zassert_equal(rem.tv_nsec, 0, "actual: %d expected: %d", rem.tv_nsec, 0);
 
-	actual_ns = k_cyc_to_ns_ceil64((now - then * selection));
+	actual_ns = k_cyc_to_ns_ceil64((now + then * selection));
```

Comment:
Actually, this is just measuring a duration, so it should be independent of selection.

This looks like it was a regression introduced in the `clock_nanosleep()` PR
```suggestion
	actual_ns = k_cyc_to_ns_ceil64(now - then);
```

Reply:
@cfriedt ,  it is related to the seletction, as last commit, the 

```
static inline int select_nanosleep(int selection, clockid_t clock_id, int flags,
				   const struct timespec *rqtp, struct timespec *rmtp)
{
	if (selection == SELECT_NANOSLEEP) {
		return nanosleep(rqtp, rmtp);
	}
	return clock_nanosleep(clock_id, flags, rqtp, rmtp);
}
```
so this code is reused for different `flags`,  and when flags == 0, we need minus the then time(uptime), but with flags==4(TIMER_ABSTIME), we do not need do this.

Reply:
Yeah, I can see now why the author decided to change the test this way. Personally, I would have preferred if the absolute time was a separate check in each case, but that would have duplicated several checks throughout the suite.

Can you adjust this change to better document the expectation here?
```diff
diff --git a/tests/posix/common/src/nanosleep.c b/tests/posix/common/src/nanosleep.c
index 5e4d3bcf88..a21309e1c8 100644
--- a/tests/posix/common/src/nanosleep.c
+++ b/tests/posix/common/src/nanosleep.c
@@ -11,8 +11,10 @@
 #include <zephyr/sys_clock.h>
 #include <zephyr/ztest.h>
 
-#define SELECT_NANOSLEEP       1
+/** @brief Test @ref clock_nanosleep */
 #define SELECT_CLOCK_NANOSLEEP 0
+/** @brief Test @ref nanosleep */
+#define SELECT_NANOSLEEP       1
 
 static inline int select_nanosleep(int selection, clockid_t clock_id, int flags,
                                   const struct timespec *rqtp, struct timespec *rmtp)
@@ -120,6 +122,22 @@ ZTEST(posix_apis, test_clock_nanosleep_errors_errno)
        zassert_equal(rem.tv_nsec, 0, "actual: %d expected: %d", rem.tv_nsec, 0);
 }
 
+/**
+ * @brief Check that a call to nanosleep has yielded executiuon for some minimum time.
+ *
+ * Check that the actual time slept is >= the total time specified by @p s (in seconds) and
+ * @p ns (in nanoseconds).
+ *
+ * @note The time specified by @p s and @p ns is assumed to be absolute (i.e. a time-point)
+ * when @p selection is set to @ref SELECT_CLOCK_NANOSLEEP. The time is assumed to be relative
+ * when @p selection is set to @ref SELECT_NANOSLEEP.
+ *
+ * @param selection Either @ref SELECT_CLOCK_NANOSLEEP or @ref SELECT_NANOSLEEP
+ * @param clock_id The clock to test (e.g. @ref CLOCK_MONOTONIC or @ref CLOCK_REALTIME)
+ * @param flags Flags to pass to @ref clock_nanosleep
+ * @param s Partial lower bound for yielded time (in seconds)
+ * @param ns Partial lower bound for yielded time (in nanoseconds)
+ */
 static void common_lower_bound_check(int selection, clockid_t clock_id, int flags, const uint32_t s,
                                     uint32_t ns)
 {
@@ -143,7 +161,18 @@ static void common_lower_bound_check(int selection, clockid_t clock_id, int flag
        zassert_equal(rem.tv_sec, 0, "actual: %d expected: %d", rem.tv_sec, 0);
        zassert_equal(rem.tv_nsec, 0, "actual: %d expected: %d", rem.tv_nsec, 0);
 
-       actual_ns = k_cyc_to_ns_ceil64((now - then * selection));
+       switch(selection) {
+               case SELECT_NANOSLEEP:
+                       /* exp_ns and actual_ns are relative (i.e. durations) */
+                       actual_ns = k_cyc_to_ns_ceil64(now - then);
+                       break;
+               case SELECT_CLOCK_NANOSLEEP:
+                       /* exp_ns and actual_ns are absolute (i.e. time-points) */
+                       actual_ns = k_cyc_to_ns_ceil64(now);
+                       break;
+               default:
+                       break;
+       }
 
        exp_ns = (uint64_t)s * NSEC_PER_SEC + ns;
        /* round up to the nearest microsecond for k_busy_wait() */
```

Reply:
I guess maybe an assert in the default selection would be good..

Reply:
updated